# Numerical Integration: Simpson's Rule
---
## _Simpson's Rule_ uses a quadratic polynomial on each subinterval of a partition to approximate the function $f(x)$ and to compute the definite integral.

<img src="https://www.emathhelp.net/images/calc/322_simpson_rule.png" width=350>

---

### This is an improvement over the trapezoid rule which approximates $f(x)$ by a straight line on each subinterval of a partition.

### The formula for Simpson's rule is
### $$S_N(f) = \frac{\Delta x}{3} \sum_{i = 1}^{N/2} (f(x_{2i - 2}) + 4f(x_{2i - 1}) + f(x_{2i}))$$

### where $N$ is an even number of sub-intervals of $[a,\,b],\,\Delta x = \frac{b - a}{N}$ and $x_i = a + i \Delta x$.

#### <span style="color:red"> Notice: In the lecture note, $N$ can be any positive integer, then $\Delta x$ will be $\Delta x = \frac{b - a}{2N}$. </span>

---
## Proof

### Pick the sub-interval $[x_{2i - 2},\,x_{2i}]$. So we can use  Lagrange polynomial interpolation to find an expression to approximate the function $f(x)$ on the sub-interval $[x_{2i - 2},\,x_{2i}]$
### $$f(x)
\approx \frac{(x - x_{2i - 1})(x - x_{2i})}{(x_{2i - 2} - x_{2i - 1})(x_{2i - 2} - x_{2i})}\,f(x_{2i - 2})
+  \frac{(x - x_{2i - 2})(x - x_{2i})}{(x_{2i - 1} - x_{2i - 2})(x_{2i - 1} - x_{2i})}\,f(x_{2i - 1})
+  \frac{(x - x_{2i - 2})(x - x_{2i - 1})}{(x_{2i} - x_{2i - 2})(x_{2i} - x_{2i - 1})}\,f(x_{2i})$$

### Thus,  the integral on the sub-interval $[x_{2i - 2},\,x_{2i}]$ is
### $$\begin{aligned}
\int_{x_{2i - 2}}^{x_{2i}} f(x)\,dx
&\approx \int_{x_{2i - 2}}^{x_{2i}} \frac{(x - x_{2i - 1})(x - x_{2i})}{(x_{2i - 2} - x_{2i - 1})(x_{2i - 2} - x_{2i})}\,f(x_{2i - 2})
+  \frac{(x - x_{2i - 2})(x - x_{2i})}{(x_{2i - 1} - x_{2i - 2})(x_{2i - 1} - x_{2i})}\,f(x_{2i - 1})\\
&+  \frac{(x - x_{2i - 2})(x - x_{2i - 1})}{(x_{2i} - x_{2i - 2})(x_{2i} - x_{2i - 1})}\,f(x_{2i})\,dx\\
&= \int_{x_{2i - 2}}^{x_{2i}} \frac{(x - x_{2i - 1})(x - x_{2i})}{2 \Delta x^2}\,f(x_{2i - 2})
+  \frac{(x - x_{2i - 2})(x - x_{2i})}{\Delta x^2}\,f(x_{2i - 1})\\
&+  \frac{(x - x_{2i - 2})(x - x_{2i - 1})}{2 \Delta x^2}\,f(x_{2i})\,dx\\
&= \frac{\Delta x}{3}\,(f(x_{2i - 2}) + 4f(x_{2i - 1}) + f(x_{2i}))
\end{aligned}$$

### where $N$ is an even number of sub-intervals of $[a,\,b],\,\Delta x = \frac{b - a}{N}$ and $x_i = a + i \Delta x$.

### Hence the Simpson's rule for the integral is
### $$\int_a^b f(x) dx \approx S_N(f) = \frac{\Delta x}{3} \sum_{i = 1}^{N/2} (f(x_{2i - 2}) + 4f(x_{2i - 1}) + f(x_{2i}))$$

---
## Prerequisites

### 1. The integrand function $f(x)$.
### 2. The interval $[a,\,b]$.
### 3. The even number $N$ where $N$ is a number of sub-intervals of  $[a,\,b]$.

---
## Implementation

### Part 0. Import necessary libraries

In [1]:
import numpy as np
import matplotlib.pyplot as plt

### Part 1. Implement Simpson's rule

In [2]:
def simpson(
    f,
    a,
    b,
    N=50
):
    
    '''
    Parameters
    ----------
    f : function
        Vectorized function of a single variable
    a , b : numbers
        Interval of integration [a,b]
    N : (even) integer
        Number of subintervals of [a,b]

    Returns
    -------
    S : float
        Approximation of the integral of f(x) from a to b using
        Simpson's rule with N subintervals of equal length.
    '''
    
    # Check whether 'N' is even
    if N % 2 == 1:
        raise ValueError("N must be an even integer.")
    
    # Compute the length of each sub-interval
    dx = (b - a) / N
    
    # Compute the equivalent points
    x = np.linspace(a, b, N+1)
    
    # Compute the value of f(x)
    y = f(x)
    
    # Compute the integal
    S = dx / 3 * np.sum(y[0:-1:2] + 4 * y[1::2] + y[2::2])
    
    return S

---
## Example

### 1. Compute the function on integrals for which we know the exact value.
### $$\int_0^1\,3x^2\,dx = 1$$

In [3]:
approximation = simpson(lambda x : 3*x**2, 0, 1, 10)
print('The approximate value is', approximation)
print('The error is', abs(1 - approximation))

The approximate value is 1.0
The error is 0.0


### 2. Let's compute another function on integrals for which we know the exact value.
### $$\int_0^{\pi/2}\,\sin x\, dx = 1$$

In [4]:
approximation = simpson(np.sin, 0, np.pi/2, 100)
print('The approximate value is', approximation)
print('The error is', abs(1 - approximation))

The approximate value is 1.000000000338236
The error is 3.382361057902017e-10


---
## Error Estimate

### We havve seen that the error in a Riemann sum is inversely proportional to the size of the partition $N$ and the trapezoid rule is inversely proportional to $N^2$.

### The errror formula in the theorem below shows that Simpson's rule is even better as the error is inversely proportional to $N^4$.

---
## Theorem

### Let $S_N (f)$ denote Simpson's rule
### $$S_N(f) = \frac{\Delta x}{3}\,\sum_{i=1}^{N/2}\,(f(x_{2i - 2}) + 4f(x_{2i - 1}) + f(x_{2i}))$$
### where $N$ is an even number of sub-intervals of $[a,\,b],\,\Delta x = \frac{b - a}{N}$ and $x_i = a + i \Delta x$.

### The <span style="color:red"> _error bound_ </span> is
### $$E_N^S(f) = \left| \int_a^b f(x)\,dx - S_N(f) \right| \leq \frac{(b - a)^5}{180N^4}\,K_4$$
### where $|f^{(4)} (x)|\,\leq\,K_4$ for all $x \in [a,b]$.

#### More detail for the proof of this theorem: [link](https://www.youtube.com/watch?v=q8jD7U0_BSs)

---
## Example

## 1. Approximate $\ln 2$

### Find a value $N$ which guarantees that Simpson's rule approximation $S_N(f)$ of the integral
### $$\int_1^2\,\frac{1}{x}\,dx$$
### satisfies $E_N^S(f) \leq 0.0001$.

### Compute
### $$f^{(4)}(x) = \frac{24}{x^5}$$
### therefore $|f^{(4)}(x)| \leq 24$ for all $x \in [1,\,2]$ and so
### $$\frac{1}{180N^4} \cdot 24 \leq 0.0001\,
\Rightarrow \, \frac{20000}{15N^4} \leq 1\,
\Rightarrow \left( \frac{20000}{15} \right)^{1/4} \leq N$$

In [5]:
# Compute the inequality
print((20000/15)**0.25, '< N')

6.042750794713537 < N


### So let's compute Simpson's rule with $N = 8$ (the smallest even integer greater than $6.04$)

In [6]:
approximation = simpson(lambda x : 1/x, 1, 2, 8)
print('The approximate value is', approximation)

The approximate value is 0.6931545306545306


### Verify that $E_N^S(f) \leq 0.0001$

In [7]:
error = abs(np.log(2) - approximation)
print('The error is', error)
print('Answer is', error <= 0.0001)

The error is 7.350094585301115e-06
Answer is True


# End